### PR TITLE
Adding support for lidar depth map conversion and attachment to the map

### PR DESCRIPTION
**Pull request by mfehr**
- Adds support for lidar to range image conversions and based on the camera model introduced in aslam_cv PR https://github.com/ethz-asl/aslam_cv2/pull/46
 - Adds option to convert lidar scans to depth maps in maplab_node. Saves a lot of space.
   - Projects to either OpenNI format (uint16_t, [mm]) or Floating point [m]
   - Optionally projects color/intensity along with the depth into color/intensity image
 - Adds option to attach point cloud to map as depth map instead of point cloud.
 - Add support to have more than one NCamera in the sensor calibration (uses the --selected... flags to determine which one is used for visual mapping). This is requried since in order to attach point clouds as depth maps, we need to add a 3DLidarCamera calibration (=another NCamera) to the calibration file. The same selection/multi-sensor option is implemented for all other sensor types as well.
 - Some minor fixes and documention for the resource importer tool
 

Significantly reduces the size of the map when attaching point clouds as depth maps, as can be seen in this example, where the full OS-1 64 beam lidar scans at 10Hz were attached:
```
9.9G	gonzen_run_7_with_pointclouds
1.5G	gonzen_run_7_with_range_image
2.2G	gonzen_run_7_range_image_and_intensity
```

### CODE REVIEW

File: backend/map-resources/include/map-resources/resource-conversion-inl.h [hunk 1]
```diff
@@ -402,6 +417,225 @@ bool convertDepthMapToPointCloud(
   return true;
 }
 
+template <typename PointCloudType>
+bool convertPointCloudToDepthMap(
+    const PointCloudType& point_cloud_C, const aslam::Camera& camera,
+    const bool use_openni_format, const bool create_range_image,
+    cv::Mat* depth_map, cv::Mat* image) {
+  CHECK(!(
+      camera.getType() == aslam::Camera::Type::kLidar3D && !create_range_image))
```
Comment (smauq) [inline backend/map-resources/include/map-resources/resource-conversion-inl.h]: Can you move the negation inside to make it a bit more readable. I.e. `camera.getType() != aslam::Camera::Type::kLidar3D || create_range_image` is more clear to me.

File: backend/map-resources/include/map-resources/resource-conversion-inl.h [hunk 2]
```diff
@@ -402,6 +417,225 @@ bool convertDepthMapToPointCloud(
...
+
+  // Initialize depth map / range image depdening on format.
+  Eigen::Matrix3Xd points_3D = Eigen::Matrix3Xd::Zero(3, num_points);
+  Eigen::VectorXd depth_values = Eigen::VectorXd::Zero(num_points);
+  Eigen::Matrix2Xd projections_2D = Eigen::Matrix2Xd::Zero(2, num_points);
+  double max_valid_depth_value = 0.;
+  if (use_openni_format) {
+    *depth_map = cv::Mat(
+        camera.imageHeight(), camera.imageWidth(), CV_16UC1, cv::Scalar(0u));
+    max_valid_depth_value =
```
Comment (smauq) [inline backend/map-resources/include/map-resources/resource-conversion-inl.h]: Your max value might be 1 less than what you want in some cases due to rounding later on, but I assume that's fine.

File: algorithms/dense-reconstruction/depth_integration/src/depth-map-integration.cc [hunk 3]
```diff
@@ -0,0 +1,349 @@
...
+                    &image)) {
+              VLOG(3) << "Found depth map with intensity information "
+                         "from the dedicated grayscale image.";
+            } else if (vi_map.getFrameResource(
+                           vertex, frame_idx, backend::ResourceType::kRawImage,
+                           &image)) {
+              VLOG(3) << "Found depth map with intensity information "
+                         "from the raw grayscale image.";
+            } else {
+              VLOG(3) << "Found depth map without intensity information.";
```
[the remark this answers is not in the document]
Comment (mfehr) [reply algorithms/dense-reconstruction/depth_integration/src/depth-map-integration.cc]: this is intended, if there is no intensity image for the depth map, we still continue, but there is just no image.

File: algorithms/dense-reconstruction/depth_integration/src/depth-map-integration.cc [hunk 4]
```diff
@@ -0,0 +1,349 @@
...
+            }
+
+            // Integrate with or without intensity information.
+            integration_function(
+                T_G_C, depth_map, image, n_camera.getCamera(frame_idx));
+            continue;
+          }
```
[the remark this answers is not in the document]
Comment (mfehr) [reply algorithms/dense-reconstruction/depth_integration/src/depth-map-integration.cc]: this is needed to define the scope for local local variables in a switch statement.

File: algorithms/dense-reconstruction/depth_integration/src/depth-map-integration.cc [hunk 5]
```diff
@@ -0,0 +1,349 @@
...
+              VLOG(3) << "Found depth map with intensity information "
+                         "from the raw grayscale image.";
+            } else {
+              VLOG(3) << "Found depth map without intensity information.";
+            }
+
+            // Integrate with or without intensity information.
+            integration_function(
+                T_G_C, depth_map, image, n_camera.getCamera(frame_idx));
+            continue;
```
[the remark this answers is not in the document]
Comment (mfehr) [reply algorithms/dense-reconstruction/depth_integration/src/depth-map-integration.cc]: makes sense

File: map-structure/vi-map/include/vi-map/sensor-utils.h [hunk 6]
```diff
@@ -14,7 +14,19 @@
 #include <sensors/pointcloud-map-sensor.h>
 #include <sensors/sensor-types.h>
 #include <sensors/wheel-odometry-sensor.h>
-#include <vi-map/sensor-manager.h>
+
```
[the remark this answers is not in the document]
Comment (mfehr) [reply map-structure/vi-map/include/vi-map/sensor-utils.h]: intentional, the last include is local (same package), so it's a separate block

File: algorithms/online_map_builders/include/online-map-builders/stream-map-builder.h [hunk 7]
```diff
@@ -199,12 +211,71 @@ void StreamMapBuilder::attachLidarMeasurement(
...
+    cv::Mat* image_ptr =
+        (FLAGS_map_builder_save_point_clouds_as_range_image_including_intensity_image)  // NOLINT
+            ? &image
+            : nullptr;
+
+    backend::convertPointCloudToDepthMap(
+        point_cloud, *lidar_depth_camera_sensor_, true /*use_openni_format*/,
+        true /*create_range_image*/, &range_image, image_ptr);
+
+    if (false) {
```
[the remark this answers is not in the document]
Comment (mfehr) [reply algorithms/online_map_builders/include/online-map-builders/stream-map-builder.h]: I'll add a flag for it, it's to show a opencv image with the depth map that the lidar scan is converted to.

**Pull request merged**